File: dsh-astrbot-relay/lib/contract.js
/**
 * AstrDsh Relay（星驿）：契约常量的机器可读副本。
 *
 * 唯一真相来源是 `docs/BRIDGE-CONTRACT.md`。
 * 本文件只是把契约里会被两侧代码同时引用的字面量集中一处，避免散落在实现里写错。
 * 修改本文件必须同步修改 docs/BRIDGE-CONTRACT.md 与 AstrBot 侧
 * `astrbot_plugin_dsh_relay/contract.py`。
 */
import { randomUUID } from 'node:crypto'

/** 契约版本。破坏性变更必须递增；/health 会返回它，IM 侧不匹配时拒绝启用。 */
export const BRIDGE_VERSION = '1'

/** 路由。契约 §3。相对基址（AstrBot 侧配置项 bridge_url）。 */
export const ROUTES = Object.freeze({
  MESSAGE: '/message',    // POST  投递用户消息（要求 Idempotency-Key）
  EVENTS: '/events',      // GET   SSE 下行通道（支持 Last-Event-ID 续传）
  APPROVAL: '/approval',  // POST  回执审批
  HEALTH: '/health',      // GET   存活与版本协商
})

/** 下行事件类型。契约 §4。 */
export const EVENT = Object.freeze({
  TURN_START: 'turn/start',
  TEXT_DELTA: 'text/delta',
  REASONING_DELTA: 'reasoning/delta',
  TOOL_CALL: 'tool/call',
  APPROVAL_REQUIRED: 'approval/required',
  APPROVAL_RESOLVED: 'approval/resolved',
  MESSAGE_FINAL: 'message/final',
  TURN_END: 'turn/end',
  HEARTBEAT: 'heartbeat',
  GAP: 'gap',
})

/** 统一错误码。契约 §8。 */
export const ERROR_CODE = Object.freeze({
  UNAUTHORIZED: 'unauthorized',
  NOT_FOUND: 'not_found',
  QUEUE_FULL: 'queue_full',
  AGENT_BUSY: 'agent_busy',
  UNSUPPORTED: 'unsupported',
  INTERNAL: 'internal',
})

/** 错误码 → HTTP 状态码。 */
export const ERROR_STATUS = Object.freeze({
  [ERROR_CODE.UNAUTHORIZED]: 401,
  [ERROR_CODE.NOT_FOUND]: 404,
  [ERROR_CODE.QUEUE_FULL]: 429,
  [ERROR_CODE.AGENT_BUSY]: 409,
  [ERROR_CODE.UNSUPPORTED]: 400,
  [ERROR_CODE.INTERNAL]: 500,
})

/**
 * 网桥暴露给 IM 的审批结论白名单。契约 §7.2 第 4 条。
 *
 * DSH 侧完整枚举是 `allowed-once | rejected | cancelled | unavailable`，
 * 但后两个是系统语义，**不允许**由 IM 用户触发，因此不在此表内。
 * 超时一律 resolve `rejected`（fail closed）。
 */
export const APPROVAL_OUTCOME = Object.freeze({
  ALLOW_ONCE: 'allowed-once',
  REJECTED: 'rejected',
})

/** IM 用户可接受的审批结论（白名单校验用）。 */
export const APPROVAL_OUTCOME_ALLOWED = Object.freeze([
  APPROVAL_OUTCOME.ALLOW_ONCE,
  APPROVAL_OUTCOME.REJECTED,
])

/** 会话轮转策略。契约 §2.3。 */
export const POLICY = Object.freeze({
  ONE_TO_ONE: 'one-to-one',
  ON_DEMAND: 'on-demand',
  DAILY: 'daily',
})

/** 该会话的 DSH sessionId 前缀，便于在 Web UI 会话列表里辨识来源。 */
export const SESSION_ID_PREFIX = 'im-'

/** 生成一个新的网桥会话 id。 */
export function newSessionId() {
  return `${SESSION_ID_PREFIX}${randomUUID()}`
}

/** 一次性的审批验证码长度（字符）。契约 §7.1。 */
export const APPROVAL_CODE_LENGTH = 4

File: astrbot_plugin_dsh_relay/contract.py
"""AstrDsh Relay（星驿）：契约常量的机器可读副本（AstrBot 侧）。

唯一真相来源是 ``docs/BRIDGE-CONTRACT.md``。
本文件与 DSH 侧 ``dsh-astrbot-relay/lib/contract.js`` 必须逐字对应；
改动任一侧都要同步契约文档与另一侧。
"""

from __future__ import annotations

#: 契约版本。破坏性变更必须递增；/health 返回它，不匹配时拒绝启用。
BRIDGE_VERSION = "1"

#: 路由。契约 §3。相对 AstrBot 侧配置项 ``bridge_url``。
ROUTE_MESSAGE = "/message"
ROUTE_EVENTS = "/events"
ROUTE_APPROVAL = "/approval"
ROUTE_HEALTH = "/health"

#: 下行事件类型。契约 §4。
EVENT_TURN_START = "turn/start"
EVENT_TEXT_DELTA = "text/delta"
EVENT_REASONING_DELTA = "reasoning/delta"
EVENT_TOOL_CALL = "tool/call"
EVENT_APPROVAL_REQUIRED = "approval/required"
EVENT_APPROVAL_RESOLVED = "approval/resolved"
EVENT_MESSAGE_FINAL = "message/final"
EVENT_TURN_END = "turn/end"
EVENT_HEARTBEAT = "heartbeat"
EVENT_GAP = "gap"

#: 未知事件类型必须被忽略（前向兼容），因此这里是"已知集合"而非白名单。
KNOWN_EVENT_TYPES = frozenset({
    EVENT_TURN_START,
    EVENT_TEXT_DELTA,
    EVENT_REASONING_DELTA,
    EVENT_TOOL_CALL,
    EVENT_APPROVAL_REQUIRED,
    EVENT_APPROVAL_RESOLVED,
    EVENT_MESSAGE_FINAL,
    EVENT_TURN_END,
    EVENT_HEARTBEAT,
    EVENT_GAP,
})

#: 统一错误码。契约 §8。
ERROR_UNAUTHORIZED = "unauthorized"
ERROR_NOT_FOUND = "not_found"
ERROR_QUEUE_FULL = "queue_full"
ERROR_AGENT_BUSY = "agent_busy"
ERROR_UNSUPPORTED = "unsupported"
ERROR_INTERNAL = "internal"

#: 可重试的错误码（含纯传输层失败）。契约 §6.2。
RETRYABLE_ERROR_CODES = frozenset({ERROR_INTERNAL})

#: IM 用户可接受的审批结论白名单。契约 §7.2 第 4 条。
#: DSH 侧完整枚举还有 ``cancelled`` / ``unavailable``，那两个是系统语义，
#: 不允许由 IM 用户触发。
APPROVAL_ALLOW_ONCE = "allowed-once"
APPROVAL_REJECTED = "rejected"
APPROVAL_OUTCOMES_ALLOWED = frozenset({APPROVAL_ALLOW_ONCE, APPROVAL_REJECTED})

#: 审批命令关键字。契约 §7.2 第 5 条：只接受明确命令，不接受"是/否"。
APPROVAL_COMMAND_APPROVE = "approve"
APPROVAL_COMMAND_REJECT = "reject"
